### PR TITLE
REBASED: Transform Speaker Encoder in a Generic Encoder and Implement Emotion Encoder training support 

### PR DESCRIPTION
**Pull request by Edresson**
PR #1316  Rebased

### CODE REVIEW

File: TTS/bin/eval_encoder.py [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+import torch
+from argparse import RawTextHelpFormatter
+
+from tqdm import tqdm
+
+from TTS.config import load_config
+from TTS.tts.datasets import load_tts_samples
+from TTS.tts.utils.speakers import SpeakerManager
+
+parser = argparse.ArgumentParser(
```
Comment (erogol) [inline TTS/bin/eval_encoder.py:11]: It would be better to make this a function so that people can call it from outside

File: TTS/bin/train_encoder.py [hunk 2]
```diff
@@ -32,142 +33,239 @@
...
     # pylint: disable=global-variable-undefined
     global meta_data_train
     global meta_data_eval
+    global train_classes
 
     ap = AudioProcessor(**c.audio)
     model = setup_speaker_encoder_model(c)
 
-    optimizer = RAdam(model.parameters(), lr=c.lr)
+    optimizer = RAdam(model.parameters(), lr=c.lr, weight_decay=c.wd)
```
Comment (erogol) [inline TTS/bin/train_encoder.py:247]: Maybe you can use get optimiser as we do it for the other models

File: TTS/encoder/models/lstm.py [hunk 3]
```diff
@@ -182,8 +182,18 @@ def batch_compute_embedding(self, x, seq_lens, num_frames=160, overlap=0.5):
     def load_checkpoint(self, config: dict, checkpoint_path: str, eval: bool = False, use_cuda: bool = False):
         state = load_fsspec(checkpoint_path, map_location=torch.device("cpu"))
         self.load_state_dict(state["model"])
+        # load the criterion for emotion classification
```
Comment (erogol) [inline TTS/encoder/models/lstm.py:185]: I would separate this into different loss classes as we do it for the TTS models

File: TTS/encoder/models/resnet.py [hunk 4]
```diff
@@ -277,8 +277,18 @@ def compute_embedding(self, x, num_frames=250, num_eval=10, return_mean=True, l2
     def load_checkpoint(self, config: dict, checkpoint_path: str, eval: bool = False, use_cuda: bool = False):
         state = load_fsspec(checkpoint_path, map_location=torch.device("cpu"))
         self.load_state_dict(state["model"])
+        # load the criterion for emotion classification
```
Comment (erogol) [inline TTS/encoder/models/resnet.py:280]: You can do the same thing as above just create different loss classes and create a loss factory and use the factory in both places.

Comment (erogol) [reply TTS/encoder/models/resnet.py:280]: Or you can implement base encoder and inherited by the subclasses

File: TTS/encoder/utils/samplers.py [hunk 5]
```diff
@@ -0,0 +1,102 @@
+import random
```
Comment (erogol) [inline TTS/encoder/utils/samplers.py:1]: Who can consider moving this to the trainers if they are also useful for other models and use cases

File: TTS/encoder/utils/samplers.py [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+from torch.utils.data.sampler import Sampler, SubsetRandomSampler
+
```
Comment (erogol) [inline TTS/encoder/utils/samplers.py:3]: It would also be good to have some tests for those classes

**Review by erogol — CHANGES_REQUESTED** [review]

File: TTS/encoder/models/lstm.py (hunk 3 above)
Comment (Edresson) [reply TTS/encoder/models/lstm.py:185]: The plan is to load this loss just for use in inference time. It is not loaded in --restore_path. For this reason, I added all these ifs. "config.map_classid_to_classname" will be different of None just during the inference. And the only Loss that we need to load is SoftmaxAngleProtoLoss because it has an nn.Dense layer.

File: TTS/bin/train_encoder.py (hunk 2 above)
Comment (Edresson) [reply TTS/bin/train_encoder.py]: Done

File: TTS/encoder/models/lstm.py (hunk 3 above)
Comment (Edresson) [reply TTS/encoder/models/lstm.py]: I create a EncoderBase Class that includes common encoder specific functions. In addition, I add a get_criterion to return the right criterion based on the config.loss.

File: TTS/encoder/models/resnet.py (hunk 4 above)
Comment (Edresson) [reply TTS/encoder/models/resnet.py]: Done, let me know if it is enough

File: TTS/encoder/utils/samplers.py (hunk 5 above)
Comment (Edresson) [reply TTS/encoder/utils/samplers.py:1]: I think it is not useful enough for now. If we need this in the future we can do that. But for now, the only use that I can see is in the encoders :).

File: TTS/bin/eval_encoder.py (hunk 1 above)
Comment (Edresson) [reply TTS/bin/eval_encoder.py]: Done

**Review by erogol — APPROVED** [review]

**Comment by ysujiang** [comment]
How to train emotion tts ? Can you give me some training details？

**Comment by erogol** [comment]
emotion TTS is still WIP. Once it is completed, we will share information in the docs

**Comment by a-ggghost** [comment]
did anything fun result from embedding prosody as well like in https://github.com/coqui-ai/TTS/commit/8a3396d9c11876c6feb4fcd20c6c4d49c3a8737a ?
